### PR TITLE
JCRVLT-301 allow to explicitly downgrade subpackages through new

### PR DESCRIPTION
**Pull request by kwin**
subpackage handling options "INSTALL_ALWAYS" and "EXTRACT_ALWAYS"

### CODE REVIEW

File: vault-core/src/main/java/org/apache/jackrabbit/vault/packaging/SubPackageHandling.java [hunk 1]
```diff
@@ -66,7 +71,19 @@
...
          * ignores the sub package completely
          */
-        IGNORE
+        IGNORE,
+        
+        /**
+         * adds and installs the package using {@link JcrPackage#install(org.apache.jackrabbit.vault.fs.io.ImportOptions)} even in case a newer version has already been installed
+         * (allows downgrades)
+         */
+        INSTALL_ALWAYS,
```
Comment (tripodsan) [inline vault-core/src/main/java/org/apache/jackrabbit/vault/packaging/SubPackageHandling.java]: how about: `FORCE_INSTALL` instead?

File: vault-core/src/main/java/org/apache/jackrabbit/vault/packaging/SubPackageHandling.java [hunk 2]
```diff
@@ -66,7 +71,19 @@
...
+        
+        /**
+         * adds and extracts the package using {@link JcrPackage#extract(org.apache.jackrabbit.vault.fs.io.ImportOptions)} even in case a newer version has already been installed
+         * (allows downgrades)
+         */
+        EXTRACT_ALWAYS
```
Comment (tripodsan) [inline vault-core/src/main/java/org/apache/jackrabbit/vault/packaging/SubPackageHandling.java]: how about `FORCE_EXTRACT` instead?

File: vault-core/src/main/java/org/apache/jackrabbit/vault/packaging/impl/JcrPackageImpl.java [hunk 3]
```diff
@@ -500,14 +496,25 @@ private void extract(Set<PackageId> processed, ImportOptions options, boolean cr
...
-                } else if (option == SubPackageHandling.Option.INSTALL) {
-                    msg = "Starting installation of subpackage " + id;
-                } else {
-                    msg = "Starting extraction of subpackage " + id;
+                String msg = null;
+                // should the package be skipped due to a newer version already installed?
+                if (option == SubPackageHandling.Option.INSTALL || option == SubPackageHandling.Option.EXTRACT) {
+                    PackageId newerPackageId = newerPackageIdPerSubPackage.get(id);
+                    if (newerPackageId != null) {
+                        msg = "Skipping installation of subpackage " + id + " due to newer installed version: '" + newerPackageId + "'";
```
Comment (tripodsan) [inline vault-core/src/main/java/org/apache/jackrabbit/vault/packaging/impl/JcrPackageImpl.java]: maybe use `String.format()` ?

File: vault-core/src/main/java/org/apache/jackrabbit/vault/packaging/impl/JcrPackageImpl.java [hunk 4]
```diff
@@ -500,14 +496,25 @@ private void extract(Set<PackageId> processed, ImportOptions options, boolean cr
...
+                        skip = true;
+                    }
+                }
+                
+                if (msg == null) {
```
Comment (tripodsan) [inline vault-core/src/main/java/org/apache/jackrabbit/vault/packaging/impl/JcrPackageImpl.java]: I would not use the `msg` variable to control the code flow. better `(!skip)`

**Comment by kwin** [comment]
@tripodsan I integrated all proposed changes in the PR. Please check again.

**Pull request closed without merging**